File: src/main/java/com/sudosoo/takeiteasy/service/PostServiceImpl.java
package com.sudosoo.takeiteasy.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.sudosoo.takeiteasy.common.service.JpaService;
import com.sudosoo.takeiteasy.dto.KafkaResponseDto;
import com.sudosoo.takeiteasy.dto.comment.CommentResponseDto;
import com.sudosoo.takeiteasy.dto.kafkaMemberValidateRequestDto;
import com.sudosoo.takeiteasy.dto.post.*;
import com.sudosoo.takeiteasy.entity.Category;
import com.sudosoo.takeiteasy.entity.Comment;
import com.sudosoo.takeiteasy.entity.Post;
import com.sudosoo.takeiteasy.kafka.KafkaProducer;
import com.sudosoo.takeiteasy.redis.RedisService;
import com.sudosoo.takeiteasy.repository.PostRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.ExecutionException;

@Service
@RequiredArgsConstructor
@Transactional
public class PostServiceImpl implements PostService , JpaService<Post,Long>{
    private final PostRepository postRepository;
    private final CategoryService categoryService;
    private final CommentService commentService;
    private final KafkaProducer kafkaProducer;
    private final ObjectMapper objectMapper;
    private final RedisService redisService;


    @Override
    public JpaRepository<Post, Long> getJpaRepository() {
        return postRepository;
    }

    @Override
    public PostResponseDto create(CreatePostRequestDto requestDto) {
        KafkaResponseDto kafkaResponseDto = null;
        try{
            String kafkaResult = kafkaProducer.replyRecord(new kafkaMemberValidateRequestDto(requestDto.getMemberId()));
            kafkaResponseDto = objectMapper.readValue(kafkaResult, KafkaResponseDto.class);

        }catch (ExecutionException | InterruptedException | IOException e ){
            e.getStackTrace();
        }
        if (kafkaResponseDto == null){
            throw new IllegalArgumentException("kafka reply error");
        }

        Post post = Post.of(requestDto);
        Category category = categoryService.getById(requestDto.getCategoryId());
        post.setMemberIdAndWriter(kafkaResponseDto.getMemberId(),kafkaResponseDto.getMemberName());
        post.setCategory(category);

        Post result = saveModel(post);

        //redis ReadRepository 데이터 삽입
        PostResponseDto responseDto = result.toResponseDto();
        redisService.saveReadValue(result.toResponseDto());
        return responseDto;
    }


    @Override
    public PostResponseDto redisTest(PostRequestDto requestDto){
        Post post = Post.testOf(requestDto);
        Category category = categoryService.getById(requestDto.getCategoryId());
        post.setCategory(category);

        Post result = saveModel(post);
        return result.toResponseDto();
    }


    @Override
    public Post getByPostId(Long postId) {
        return findModelById(postId);
    }

    @Override
    @Transactional(readOnly = true)
    public PostDetailResponseDto getPostDetailByPostId(Long postId, Pageable pageRequest) {
        Post post = findModelById(postId);
        post.incrementViewCount();

        Page<Comment> comments = commentService.getCommentsByPostId(postId,pageRequest);
        //TODO MemberSetting 각 코멘트의 유저 이름을 찾아와서 넣어주기
        List<CommentResponseDto> responseCommentDtos = comments.stream().map(Comment::toResponseDto).toList();
        return post.toDetailDto(new PageImpl<>(responseCommentDtos));
    }

    @Override
    @Transactional(readOnly = true)
    public List<PostResponseDto> getAll() {
        List<Post> posts = postRepository.findAll();
        return posts.stream().map(Post::toResponseDto).toList();
    }

    @Override
    public List<PostTitleOnlyResponseDto> getPagination(Pageable pageable) {
        return postRepository.findAll(pageable).map(Post::toTitleOnlyDto).toList();
    }


    @Override
    public Post createBatchPosts(int count){
        return Post.builder()
                .title("title"+count)
                .content("content"+count)
                .memberId((long) count)
                .writerName("member"+count)
                .build();
    }

}

File: takeItEasy-application/src/main/kotlin/com/sudosoo/takeItEasy/application/service/PostServiceImpl.kt
package com.sudosoo.takeItEasy.application.service

import com.fasterxml.jackson.databind.ObjectMapper
import com.sudosoo.takeItEasy.domain.repository.common.BaseRepository
import com.sudosoo.takeItEasy.application.common.jpa.JpaService
import com.sudosoo.takeItEasy.application.common.specification.JpaSpecificService
import com.sudosoo.takeItEasy.application.dto.comment.CommentResponseDto
import com.sudosoo.takeItEasy.application.dto.kafka.KafkaResponseDto
import com.sudosoo.takeItEasy.application.dto.kafka.kafkaMemberValidateRequestDto
import com.sudosoo.takeItEasy.application.dto.post.*
import com.sudosoo.takeItEasy.application.dto.post.specification.PostSpec
import com.sudosoo.takeItEasy.domain.entity.Comment
import com.sudosoo.takeItEasy.domain.entity.Post
import com.sudosoo.takeItEasy.application.kafka.KafkaProducer
import com.sudosoo.takeItEasy.application.redis.RedisService
import com.sudosoo.takeItEasy.domain.entity.EsPost
import com.sudosoo.takeItEasy.domain.entity.Event
//import com.sudosoo.takeItEasy.domain.entity.
import com.sudosoo.takeItEasy.domain.repository.CommentRepository
import com.sudosoo.takeItEasy.domain.repository.PostElasticRepository
//import com.sudosoo.takeItEasy.domain.repository.PostElasticRepository
import com.sudosoo.takeItEasy.domain.repository.PostRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.elasticsearch.core.query.Query.findAll
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.io.IOException
import java.util.concurrent.ExecutionException

@Service
@Transactional
class PostServiceImpl(
    private val postRepository: PostRepository,
    private val categoryService: CategoryService,
    private val commentRepository: CommentRepository,
    private val kafkaProducer: KafkaProducer,
    private val redisService: RedisService
) : PostService , JpaService<Post, Long>, JpaSpecificService<Post, Long>{
    override var jpaRepository: BaseRepository<Post, Long> = postRepository
    override val jpaSpecRepository: BaseRepository<Post, Long> = postRepository
    val objectMapper = ObjectMapper()

    override fun defaultCreate(requestDto: CreatePostRequestDto) {
        val post = Post(requestDto.title, requestDto.content)
        val category = categoryService.getById(requestDto.categoryId)
        post.category = category
        save(post)
    }

    override fun create(requestDto: CreatePostRequestDto): TestPostResponseDto {
        var kafkaResponseDto: KafkaResponseDto? = null
        try {
            val kafkaResult = kafkaProducer.replyRecord(kafkaMemberValidateRequestDto(requestDto.memberId))
            kafkaResponseDto = objectMapper.readValue(kafkaResult, KafkaResponseDto::class.java)
        } catch (e: ExecutionException ) {
            e.stackTrace
        } catch (e: InterruptedException) {
            e.stackTrace
        } catch (e: IOException) {
            e.stackTrace
        }
        requireNotNull(kafkaResponseDto) { "kafka reply error" }

        val post = Post(requestDto.title, requestDto.content)
        val category = categoryService.getById(requestDto.categoryId)
        post.setMemberIdAndWriter(kafkaResponseDto.memberId, kafkaResponseDto.memberName)
        post.category = category
        val result: Post = postRepository.save(post)

        //redis ReadRepository 데이터 삽입
        val responseDto = TestPostResponseDto(result)
        redisService.saveReadValue(responseDto)
        return responseDto
    }

    override fun getByPostId(postId: Long): Post {
        return findById(postId) }

    @Transactional(readOnly = true)
    override fun getPostDetailByPostId(postId: Long, pageRequest: Pageable): PostDetailResponseDto {
        val post: Post = postRepository.findById(postId)
            .orElseThrow{ IllegalArgumentException("해당 게시물이 존재 하지 않습니다.") }
        post.incrementViewCount()

        val comments: Page<Comment> = commentRepository.findCommentsByPostId(postId, pageRequest)
        val responseCommentDtos : MutableList<CommentResponseDto> =
            comments.stream().map{ o -> CommentResponseDto(o) }.toList()

        return PostDetailResponseDto(post,responseCommentDtos)
    }

    override fun getPaginationPost(pageRequest: Pageable): Page<PostTitleOnlyResponseDto> {
        val posts: Page<Post> = postRepository.findAll(pageRequest)
        val titleOnlyPost = posts.stream().map{ o -> PostTitleOnlyResponseDto(o)}.toList()
        return PageImpl(titleOnlyPost, pageRequest, posts.totalElements)
    }

    override fun createBatchPosts(count: Int): Post {
        return Post("Title$count",  "content$count",  count.toLong())
    }

    override fun softDeletePost(postId: Long) {
        val post = findById(postId)
        post.delete()
        save(post)
    }


}